File: MyQueue.js
/**
 * JavaScript implementation of `queue with two stacks` for
 * comparison.  Actually doesn't look much better, but the
 * enqueue multiple elements is cool IMO
 */
var MyQueue;

(MyQueue = function () {
    // Initialize incoming and outgoing queues
    this.in = [];
    this.out = [];
}).prototype = {
    enqueue: function () {
        // .concat merges two arrays together (appends all elements from
        // the source array to the target).  This is an idempotent action,
        // so it has to be written back to the original array (hence the `=`).
        //
        // The `.slice.call` is done to convert the special `arguments` object
        // to an array.  That is hyper-JavaScript specific.
        this.in = this.in.concat(Array.prototype.slice.call(arguments));
    },
    dequeue: function () {
        // Outgoing queue is not empty, so pop it.  JavaScript Arrays implement
        // stack operations.
        if (this.out.length) {
            return this.out.pop();
        }

        // In reality, this should be:
        // while (var x = this.in.pop()) { this.out.push(x)' }
        //
        // However, `.reverse` flips the incoming stack just as well.
        //
        // The incoming stack would be cleared by all the pop operations, so
        // we simply *clear* it by rewriting an emtpy array
        this.out = this.in.reverse();
        this.in = [];

        return this.out.pop();
    }
}

var x,
    q = new MyQueue;
q.enqueue("one", "two", "three");
console.log(q.dequeue());
q.enqueue("four");
console.log(q.dequeue());
q.enqueue("five");
while (x = q.dequeue()) {
    console.log(x);
}

File: MyQueue.java
// Hard to keep comments serious here (see MyStack), but this file will
// be a lot more rigid and serious

// Import Java's stack implementation.  I could implement `YourStack`,
// but this would require using some alternative to native arrays such as
// `ArrayList`.  Since those have to be imported anyway, I might as well import
// `Stack`.  Why the fuck should you write a Java program without importing
// anything?  Who programs without reading any documentation?
import java.util.Stack;

public class MyQueue<T> {
  // Inbox
  private Stack<T> in = new Stack<T>();

  // Outbox
  private Stack<T> out = new Stack<T>();

  // Add FIFO element
  public void enqueue(T elem) {
    in.push(elem);
  }

  // Queues are "FIFO..." first-in-first-out.  This means that the *first*
  // element to be enqueued will be the first returned on dequeue, even if it
  // were enqueued long ago.
  //
  // In order to do this, what we actually do is flip the entire `in` stack
  // by repeatedly popping it and pushing it to the `out` stack.
  //
  // Imagine that we've only enqueued so far ... On the *first* dequeue
  // it would look like this:
  //
  //     c         _            _         _           _          _            _         a
  //     b         _     ->     b         _     ->    _          b     ->     _         b
  //     a         _            a         c           a          c            _         c
  // IN STACK  OUT STACK    IN STACK  OUT STACK    IN STACK  OUT STACK    IN STACK  OUT STACK
  //
  // The `in` stack has been flipped on its head
  //
  // ***Remember*** that *stacks* go `LIFO`, that is *last-in-first-out*, so
  // since we see that `a` as at the *top* of the `out` stack, it will be
  // the first to be popped
  //
  // As far as our Queue knows, that means that `a` is the first element to
  // be dequeued.  In fact, when dequeue is called, `a` is popped immediately,
  // so we see this:
  //
  //     _
  //     _         b
  //     _         c
  // IN STACK   OUT STACK
  //
  // If we try to dequeue again (in fact, *any* time we try dequeue while
  // `out` is not empty) we simply pop the `out` stack.  Once `out` is empty
  // again, go back to line 30.
  public T dequeue() {

    // As (hopefully) made clear above, continue to pop from the outgoing
    // stack which is the flipped incoming stack and thus has switched
    // the LIFO `Stack` to the FIFO `Queue` by dropping it on its head
    //
    // We only need to pop once per `dequeue` since `dequeue` only returns
    // one element
    if (!out.empty()) {
      // The operation ends here.  If `out` was not empty, *stop reading
      // all the other comments*!
      return out.pop();
    }

    // The outgoing stack is empty, so there is work to be done.  We have
    // to flip the incoming stack on its head *into* the outgoing stack.
    if (!in.empty()) {
      // The complete flip, top-to-bottom
      while (!in.empty()) {
        out.push(in.pop());
      }
      // Our goal is to return the very first element pushed/enqueued
      // via `dequeue`.  This will wind up at the top of the `out` stack
      // because it is at the bottom of the `in` stack, which has been flipped
      return out.pop();
    }

    // Both stacks are empty
    return null;
  }

  // This is just done for some sugar in the sample ... Make it easy to
  // dequeue in a loop
  public boolean empty() {
    return in.empty() && out.empty();
  }

  public static void main(String[] args) {
    MyQueue<String> q = new MyQueue<String>();
    q.enqueue("one");
    q.enqueue("two");
    q.enqueue("three");
    System.out.println(q.dequeue());
    q.enqueue("four");
    System.out.println(q.dequeue());
    q.enqueue("five");
    while (!q.empty()) {
      System.out.println(q.dequeue());
    }
  }
}
